src/migrations/rebase.rs

```diff
@@ -0,0 +1,286 @@
+            // add 1 to index since we want the migrations not apart of source
+            return Ok(RebaseMigrations {
+                base_migrations: source_migrations,
+                source_migrations: new_source_migrations,
+                target_migrations: target_migrations.split_off(index + 1)
+            })
+        }
+    }
+
+    Ok(RebaseMigrations {
+        base_migrations: IndexMap::new(),
+        source_migrations,
+        target_migrations
+    })
+}
+
+async fn rebase_migration_ids(context: &Context) -> anyhow::Result<()> {
+    let migrations = migration::read_all(context, false).await?;
+    let mut changed_ids: HashMap<String, String> = HashMap::new();
+
+    for (id, mut migration_file) in migrations {
+        let mut migration_text = fs::read_to_string(&migration_file.path)?;
+
+        // its important to change parent before the main ID, since parent effects the hash id of the migration
+        if changed_ids.contains_key(&migration_file.data.parent_id) {
+            let new_parent_id = changed_ids.get(&migration_file.data.parent_id).unwrap();
+            migration_text = migration_file.data.replace_parent_id(&migration_text, new_parent_id);
+            migration_file.data.parent_id = new_parent_id.clone(); // change for hash computation below
+        }
+
+        let expected_id = migration_file.data.expected_id(&migration_text)?;
+
+        if id != expected_id {
+            migration_text = migration_file.data.replace_id(&migration_text, &expected_id);
+            migration_file.data.id = expected_id.clone();
+            changed_ids.insert(id, expected_id);
+        }
+
+        fs::write(&migration_file.path, migration_text)?;
+    }
+
+    Ok(())
+}
+
+pub async fn do_rebase(connection: &mut Connection, context: &Context, rebase_migrations: RebaseMigrations) -> anyhow::Result<()> {
+    let temp_dir = tempfile::tempdir()?;
+    let temp_ctx = Context {
+        schema_dir: temp_dir.path().to_path_buf(),
+        edgedb_version: None,
+        quiet: false,
+    };
+
+    // write all the migrations to disc.
+    let flattened_migrations = rebase_migrations.flatten()?;
+    for migration in &flattened_migrations {
+        create::write_migration(&temp_ctx, migration, false).await?;
+    }
+
+    // update the IDs of the migrations for the rebase, since we're changing the history the IDs can be invalid so we
+    // need to update them.
+    rebase_migration_ids(&temp_ctx).await?;
+
+    // remove the old migrations
+    for old in migration::read_names(context).await? {
+        fs::remove_file(old)?;
+    }
+
+    // move the new migrations from temp to schema dir
+    let mut last: Option<&RebaseMigrationKind> = None;
+    for (index, from) in migration::read_names(&temp_ctx).await?.iter().enumerate() {
+        let rebase_migration = flattened_migrations.get(index);
+        let to = context
+            .schema_dir
+            .join("migrations")
+            .join(from.file_name().expect(""));
```

Comment:
`.expect("")` will produce confusing panics.

```suggestion
            .join(from.file_name().unwrap());
```

I think that it is safe to assume that `from` will have a file name, because this temporary dir will only contain files that we have written, which will never be a directory.